engine/src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala

```diff
@@ -181,15 +180,38 @@ class SlickDataAccess(databaseConfig: Config) extends DataAccess {
     val actionThreadPoolSize = databaseConfig.getIntOr("actionThreadPoolSize", dbNumThreads) min dbMaximumPoolSize
     Executors.newFixedThreadPool(actionThreadPoolSize)
   }
+
   private val actionExecutionContext: ExecutionContext = ExecutionContext.fromExecutor(
-    actionThreadPool, database.executor.executionContext.reportFailure)
+    actionThreadPool, database.executor.executionContext.reportFailure
+  )
+
+  private lazy val useSlickUpserts = dataAccess.driver.capabilities.contains(JdbcProfile.capabilities.insertOrUpdate)
 
   override def close(): Unit = {
     actionThreadPool.shutdown()
     database.close()
   }
 
-  // Run action with an outer transaction
+  private def runTransactionWithRollbackRetry[R](action: DBIOAction[R, _ <: NoStream, _ <: Effect], actorSystem: ActorSystem): Future[R] = {
+    def retry[T](f: => Future[T], delay: FiniteDuration, retries: Int): Future[T] = {
+      implicit val ec = actorSystem.dispatcher
+      f recoverWith {
+        case e: MySQLTransactionRollbackException if retries > 0 =>
+          /** https://dev.mysql.com/doc/refman/5.5/en/innodb-deadlocks.html
+            *
+            * "Normally, you must write your applications so that they are always
+            *  prepared to re-issue a transaction if it gets rolled back because of a deadlock."
+            */
+          println(s"MySQL Transaction rollback detected.  Retrying in $delay ($retries retries remaining)")
+          after(delay, actorSystem.scheduler)(retry(f, delay, retries - 1))
+      }
+    }
+
+    def queryAttempt = database.run(action.transactionally)
+
+    Future(Await.result(retry(queryAttempt, delay=200 millis, retries=10), Duration.Inf))(actionExecutionContext)
```

Comment:
Yeah, I see what you mean.  The code does do `Await.result(retry(queryAttempt, delay=200 millis, retries=10))` so it does seem that in the worst case it will be waiting 2 seconds if all 10 retries happen.  And that whole time it's waiting, other queries are not running in that slot.  And actually, if my understanding is correct, then the retries as I have them are really no better than `Thread.sleep()`.

I guess an alternative would be to yield that slot to other queries if a deadlock occurs?  Though I'm not sure if that fixes anything because wouldn't the caller of `upsertExecutionInfo()` just have to wait longer for the results?

I don't know...  I feel like I'm not understanding something.  Maybe this is good to talk over with @kshakir